pkg/provider/docker/config.go

```diff
@@ -345,7 +345,17 @@ func (p Provider) getIPAddress(ctx context.Context, container dockerData) string
 			logger.Warnf("Unable to get IP address for container %s : Failed to inspect container ID %s, error: %s", container.Name, connectedContainer, err)
 			return ""
 		}
-		return p.getIPAddress(ctx, parseContainer(containerInspected))
+
+		// Check connected container for traefik.docker.network, falling back to
+		// the network specified on the current container.
+		containerParsed := parseContainer(containerInspected)
+		extraConf, _ := p.getConfiguration(containerParsed)
```

Comment:
I chose not to handle the error here, because this check previously didn't happen at all, but if there's a log line worth outputting here, let me know how you'd like that handled.

Reply:
yeah, let's handle it like in the rest of the code please:

```
		if err != nil {
			logger.Warnf("Unable to get IP address for container %s : failed to get extra configuration for container %s: %s", container.Name, inspected.Name, err)
			return ""
		}
```

Reply:
@mpl Hmm... It's different to the warning on L345, where we have failed to get any information about that connected container:

https://github.com/bjeanes/traefik/blob/container-networking-use-specified-container/pkg/provider/docker/config.go#L343-L347

At the site where I skipped error handling, the connected container has been fetched and identified, but parsing the extra configuration might have failed. Essentially, we can see the container, but we may not be able to parse any `traefik.docker.*` labels on it.

Is an early return the right direction? It seems extreme to me, because we may still get an IP address from that container, it just might not be the one corresponding to the specified network. There is already a warning that will be raised in that case though. Essentially, it would be the existing behaviour prior to this patch.

A warning here on L352 definitely seems pragmatic, but the code snippet you've added in your comment doesn't seem like a home run to me.

I could be missing something here, as I'm less familiar with this code, of course... 

Let me know if I've got this wrong, and I'll make the change!

Reply:
What would be the point of getting an IP address which has little to do with how the user intended to configure things? it seems to me that it might give them a fake sense of comfort, and might make them miss the fact that they mis-configured something. If we fail hard and return no IP address, there's a better chance the misconfiguration will get noticed.

That is, unless you see a case where p.getConfiguration would fail for another reason than misconfiguration, and in which case, it might make sense to try and cope with it.

Reply:
@mpl it's a breaking change, is all.

Reply:
As far as I can tell, though, this is the behaviour in Traefik for a more conventional setup. When the specified network can't be found, it assigns a non-deterministic IP from one of the networks on the container.

https://github.com/containous/traefik/blob/6b1158235e145635d79d251ac1d8a3865258e329/pkg/provider/docker/config.go#L313-L323

In that case, if an IP is found from that network, it returns early with the IP. If it is _not_ found, it warns and continues.

This PR is just delegating that same logic to the container which owns the network stack. This section of my PR simply makes sure the referenced container struct has the necessary metadata so that the existing logic can do the right thing.

If we want to fail when the specified network is non-existent, shouldn't it be done in the code I just referenced, so that it is consistent across all setups?

@mpl If you still feel, given the above, that this should be treated differently and fail early, then just confirm again and I will happily do as advised. Cheers.

Reply:
Bump!

Reply:
I understand what you're saying, and I understand the argument of consistency (i.e. if at one point we fail to find the network, then we should always leave it to the final loop to pick an address at random in one of our networks).

However, if we wanted to be really consistent everywhere, then I see no reason why we should keep the returns at L339 and L346. These errors too could be considered as non-critical, and we would also go YOLO on them and let the final for loop pick an address in that case. And same for L332 too actually. Right?

And I think the crux of it really is that when we're in IsHost or IsContainer mode, we try to respect the user's intent when they configured that. We assume we knew what they wanted and they just made a mistake while doing it, so we're not sneakily picking something else at random against their intent.

And so that's why it might be ok end up in the final loop at L361 only if we are in the error case of L321, even though it does not look consistent at first sight. I think we should either go that way, or go "full consistent" everywhere else and change all the other returns as well (L332, L339, L346) but I'm very wary of that last proposition, as it would probably break a number of things.

Reply:
Thanks for the response and clarity on your position. I'll address the feedback ASAP

Reply:
Sorry, I re-read that and realised it _could_ come off as passive aggressive. I hear you, and that logic makes total sense so I am very onboard. I'll try to fix this up today.

Reply:
Np, you were the right amount of insistent as far as I am concerned. Enough to guilt me into coming back to the PR regularly, but not enough to piss me off. ;)